File: js/src/trees/isSame.js
const isSame = (root) => {
	const dfs = (a, b) => {
		if (!a && !b) return true;
		if (!a || !b) return false;
		if (a.val !== b.val) return false;

		const left = dfs(a.left, b.left);
		const right = dfs(b, left, b.left);

		return left && right;
	};

	return dfs(a, b);
};

File: python/trees/isSame.py
from typing import Optional


class TreeNode:
    def __init__(
        self,
        val=0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False

        left, right = self.isSameTree(p.left, q.left), self.isSameTree(p.right, q.right)

        return left and right
